crates/ibc/src/core/ics02_client/consensus_state.rs

```diff
@@ -1,87 +1,32 @@
 
-/// Abstract of consensus state information used by the validity predicate
-/// to verify new commits & state roots.
+/// Derive macro that implements [`ConsensusState`] for enums containing
+/// variants that implement [`ConsensusState`]
+pub use ibc_derive::ConsensusState;
+
+/// Defines methods that all `ConsensusState`s should provide.
 ///
-/// Users are not expected to implement sealed::ErasedPartialEqConsensusState.
-/// Effectively, that trait bound mandates implementors to derive PartialEq,
-/// after which our blanket implementation will implement
-/// `ErasedPartialEqConsensusState` for their type.
-pub trait ConsensusState:
-    AsAny
-    + sealed::ErasedPartialEqConsensusState
-    + DynClone
-    + ErasedSerialize
-    + ErasedProtobuf<Any, Error = ClientError>
-    + core::fmt::Debug
-    + Send
-    + Sync
-{
+/// One can think of a "consensus state" as a pruned header, to be stored on chain. In other words,
+/// a consensus state only contains the header's information needed by IBC message handlers.
+pub trait ConsensusState: Send + Sync {
     /// Commitment root of the consensus state, which is used for key-value pair verification.
     fn root(&self) -> &CommitmentRoot;
 
     /// The timestamp of the consensus state
     fn timestamp(&self) -> Timestamp;
 
-    /// Convert into a boxed trait object
-    fn into_box(self) -> Box<dyn ConsensusState>
-    where
-        Self: Sized,
-    {
-        Box::new(self)
-    }
-}
-
-// Implements `Clone` for `Box<dyn ConsensusState>`
-dyn_clone::clone_trait_object!(ConsensusState);
-
-// Implements `serde::Serialize` for all types that have ConsensusState as supertrait
-#[cfg(feature = "serde")]
-erased_serde::serialize_trait_object!(ConsensusState);
-
-impl PartialEq for dyn ConsensusState {
-    fn eq(&self, other: &Self) -> bool {
-        self.eq_consensus_state(other)
-    }
-}
-
-// see https://github.com/rust-lang/rust/issues/31740
-impl PartialEq<&Self> for Box<dyn ConsensusState> {
-    fn eq(&self, other: &&Self) -> bool {
-        self.eq_consensus_state(other.as_ref())
-    }
-}
-
-mod sealed {
-    use super::*;
-
-    pub trait ErasedPartialEqConsensusState {
-        fn eq_consensus_state(&self, other: &dyn ConsensusState) -> bool;
-    }
-
-    impl<CS> ErasedPartialEqConsensusState for CS
-    where
-        CS: ConsensusState + PartialEq,
-    {
-        fn eq_consensus_state(&self, other: &dyn ConsensusState) -> bool {
-            other
-                .as_any()
-                .downcast_ref::<CS>()
-                .map_or(false, |h| self == h)
-        }
-    }
+    /// Serializes the `ConsensusState`. This is expected to be implemented as
+    /// first converting to the raw type (i.e. the protobuf definition), and then
+    /// serializing that.
+    ///
+    /// Note that the `Protobuf` trait in `tendermint-proto` provides convenience methods
+    /// to do this automatically.
+    fn encode_vec(&self) -> Result<Vec<u8>, tendermint_proto::Error>;
```

Comment:
The `Protobuf` trait in ibc-proto-rs is trait object safe, which it trades off with being more complex, such as having `Self: erased::TryFrom<Raw>` instead of just `Self: TryFrom<Raw>`. Now that we no longer need trait object safety, we no longer need this additional complexity.

We can just bring the same changes and make `tendermint_proto::Protobuf::encode_vec()` return `Vec<u8>`?

Reply:
Hmm. We can do so but also was thinking of not being dependent on tendermint-specific things in the "core" module.

Reply:
Indeed that would be desirable, but we should do this in another PR